Issue/#114 cli editor support

timothysparg opened the pull request:
Resolves #114

timothysparg commented:
Is there a better way to structure this for testing?

ddgenome commented on lib/kubeCrypt.ts:
```diff
@@ -109,3 +122,59 @@ function wrapLiteral(literal: string, prop: string): DeepPartial<k8s.V1Secret> {
+    const tmpFile = tempy.file();
+    try {
+        fs.writeFileSync(tmpFile, fileText);
+        child_process.spawnSync(process.env.EDITOR || "vi", [tmpFile], {
```
We prefer spawnPromise from `@atomist/sdm` because it has better cross-platform support that the node provided spawn.

ddgenome commented on package.json:
```diff
@@ -57,6 +57,7 @@
     "read-pkg-up": "^6.0.0",
     "request": "^2.88.0",
     "source-map-support": "^0.5.12",
+    "tempy": "^0.3.0",
```
See about comment about tmp-promise.

ddgenome commented on lib/kubeCrypt.ts:
```diff
@@ -109,3 +122,59 @@ function wrapLiteral(literal: string, prop: string): DeepPartial<k8s.V1Secret> {
+            outputSecret = await yaml.safeLoad(secretText);
+            errorMessage = "";
+        } catch (e) {
+            errorMessage = `# ${e.message.replace(/\n/gm, "\n#")} \n`;
```
```suggestion
            errorMessage = `# ${e.message.replace(/\n/gm, "\n# ")} \n`;
```

timothysparg commented on lib/kubeCrypt.ts:
```diff
@@ -20,9 +20,12 @@ import {
 import * as inquirer from "inquirer";
 import * as yaml from "js-yaml";
+import * as _ from "lodash";
+import * as tempy from "tempy";
```
Ah I didn't see that. Saw [tempfile](https://www.npmjs.com/package/tempfile) being used in some of the other repos so it seemed natural to use tempy.

Will update to use tmp-promise

ddgenome commented:
Regarding test, for starters you could set the editor to `true` and ensure that the round trip leaves the content unmodified. You could also test using sed, if you are on a system likely to have it.

timothysparg commented:
> It seems the edit capability is neither part of kube-encrypt or kube-decrypt, since what you really want is a round trip, decrypt-edit-encrypt. Maybe it should be its own command, kube-edit.

ok, let me rework as it's own command and see how that looks.

timothysparg commented:
@ddgenome do you mind having a look and seeing if this is heading in the right direction?

timothysparg commented on lib/kubeEdit.ts:
```diff
@@ -0,0 +1,125 @@
+
+        await spawnPromise(process.env.EDITOR || "vi", [path], {
+            stdio: "inherit",
+        });
```
This doesn't work as expected yet - vim doesn't seem to be in the same process (closes prematurely)

ddgenome replied:
What is the exact error, if any, that is printed? Perhaps there is a node module that already does something like this?

-   https://www.npmjs.com/package/editor
-   https://www.npmjs.com/package/open-editor
-   https://www.npmjs.com/package/open-in-editor

timothysparg replied:
Vim just opens and closes, with the following in the console

> Vim: Error reading input, exiting...

In light of your other comment about the handling for other editors, it would most probably make sense to explore one of the those modules that specifically deals with this.

ddgenome commented on index.ts:
```diff
@@ -283,6 +284,34 @@ function setupYargs(yargBuilder: yb.YargBuilder): void {
+            describe: "Edit Kubernetes secret data values from secret spec file",
+            type: "string",
+        }, {
+            parameterName: "literal",
```
Does it make sense to have a literal option?

timothysparg replied:
not anymore, will remove

ddgenome commented on index.ts:
```diff
@@ -283,6 +284,34 @@ function setupYargs(yargBuilder: yb.YargBuilder): void {
+        command: "kube-edit",
+        describe: "Decrypts a secret and opens it in an editor and then outputs the encrypted result",
+        parameters: [{
+            parameterName: "file",
```
Does it make sense to have file be an option? Perhaps it should be a required command-line argument.

timothysparg replied:
makes sense, will update

ddgenome commented on index.ts:
```diff
@@ -283,6 +284,34 @@ function setupYargs(yargBuilder: yb.YargBuilder): void {
+            describe: "Key to use to encrypt & decrypt secret data values",
+            type: "string",
+        }, {
+            parameterName: "base64",
```
We are basically talking about editing a Kubernetes secret here, so I think this should probably always decode/encode Base64.

timothysparg replied:
agreed

ddgenome commented on lib/kubeEdit.ts:
```diff
@@ -0,0 +1,125 @@
+    await tmp.withFile(async ({path, fd}) => {
+        await fs.writeFile(path, fileText);
+
+        await spawnPromise(process.env.EDITOR || "vi", [path], {
```
If someone sets their EDITOR environment variable to something like "emacs -nw", this will not work. You'll need to split EDITOR on whitespace and set the command and args accordingly.

timothysparg commented:
> It looks like a good start. I couple guiding questions that may make the implementation easier.

With regard to the questions about the cli options, would it then make sense to introduce a `--save` operator, or would we leave that up to the user ie `kube-edit --file=blah --secret-key=🔑  >> blah`

ddgenome commented:
When would you not want to save? I am thinking the `kube-edit` command takes an encoded and optionally encrypted secret spec, optionally decrypts, decodes, writes that to a temp file, opens an editor on that file, waits for the editor to exit, reads the file back in, encodes, optionally encrypts, and writes that to the original file.

This pattern is useful for GitOps users, so you can always recover the original contents of the secret spec from the Git history.

timothysparg commented:
ah ok, I missed that initially - thought we would just be writing to std out.

timothysparg commented:
@ddgenome mind having a look when you have a chance?

ddgenome commented on index.ts:
```diff
@@ -283,6 +284,24 @@ function setupYargs(yargBuilder: yb.YargBuilder): void {
+            parameterName: "file",
+            describe: "Edit Kubernetes secret data values from secret spec file",
+            type: "string",
+            required: true,
```
Sorry I was not clearer on this before. I think having a required option is an oxymoron. If people have to provide it, why make them type `--file=`? Just grab it from `argv._` and complain if it is not there.

timothysparg replied:
I've changed this to use a `positional` argument, which seems to work pretty well

ddgenome commented on lib/kubeCrypt.ts:
```diff
@@ -51,46 +55,25 @@ export interface KubeCryptOptions {
+    try {
+        secret = await handleSecretParameter(opts);
+    } catch (e) {
+        print.error(`Failed to load secret spec from file '${opts.file}': ${e.message}`);
```
This error message is not accurate anymore. Perhaps all the try/catch should be in the `handleSecretParameter` functions.

timothysparg replied:
That try/catch is used more for exit handling than anything else, would then need to move exit (code) handling into the `handleSecretParameter` function, which feels a little inconsistent as the `crypt` method also uses the same pattern for exit code handling

ddgenome replied:
Since only the reading of the file can throw an error, it seems to make sense to handle that issue there and use a different mechanism, e.g., returning `undefined`, to signal a failure of the function that then results in exit status handling.

timothysparg replied:
ah ok, that makes sense. Will update accordingly

ddgenome commented on lib/kubeEdit.ts:
```diff
@@ -0,0 +1,123 @@
+}
+
+function crypAction(opts: kubeEditOptions, KubeCryptAction: KubeCryptActions): KubeCryptOptions {
+    return { ...opts, ...{ action: KubeCryptAction } };
```
```suggestion
    return { ...opts, action: KubeCryptAction };
```

ddgenome commented on test/kubeUtils.test.ts:
```diff
@@ -18,9 +18,9 @@ import * as k8s from "@kubernetes/client-node";
-    cryptEncode,
-} from "../lib/kubeCrypt";
+base64,
+crypt,
```
Indentation

ddgenome commented on test/kubeUtils.test.ts:
```diff
@@ -110,7 +110,8 @@ describe("kubeCrypt", () => {
 
-            const encodedSecret = await cryptEncode(secret, "super-secret-key", true, true);
+            let encodedSecret = base64(secret, "encode");
+            encodedSecret = await crypt(encodedSecret, { action: "encrypt", secretKey: "super-secret-key" });
```
We prefer using two `const` rather than reassigning a `let`.

ddgenome commented on test/kubeUtils.test.ts:
```diff
@@ -158,7 +159,8 @@ describe("kubeCrypt", () => {
 
-            const encodedSecret = await cryptEncode(secret, "super-secret-key", false, true);
+            let encodedSecret = await crypt(secret, { action: "decrypt", secretKey: "super-secret-key" });
+            encodedSecret = base64(encodedSecret, "decode");
```
Two `const` again.

timothysparg commented:
@ddgenome do you mind reviewing again?

ddgenome commented on lib/kubeCrypt.ts:
```diff
@@ -51,46 +55,25 @@ export interface KubeCryptOptions {
+    try {
+        secret = await handleSecretParameter(opts);
+    } catch (e) {
+        print.error(`Failed to load secret spec from ${opts.file ? `'${opts.file}'` : "--file or --literal"}: ${e.message}`);
```
If `opts.file` is not truthy, why would you say "--file or --literal"?

timothysparg commented:
@ddgenome 

I've updated the PR with the latest changes:
-  handleSecretParameter now returns undefined and internally handles the try/catch
- Added [`conflicts: "literal"`](https://github.com/atomist/cli/pull/121/files#diff-ed009b6b86b017532ef0489c77de5100R267) and [`conflicts: "file"`](https://github.com/atomist/cli/pull/121/files#diff-ed009b6b86b017532ef0489c77de5100R272) cli restrictions for kube-encrypt/decrypt
- Added [`test/kubeDecrypt.cli.test.ts`](https://github.com/atomist/cli/pull/121/files#diff-21d44671e904b7480cd2f0c3acb77d3f) and [`test/kubeEncrypt.cli.test.ts`](https://github.com/atomist/cli/pull/121/files#diff-c7d64d62a17d8c1c6b534e88044e5752)


I would really like to add some interactive cli testing ( the [`suppose`](url) package looks particularly promising for this ) - but I think that is for another pull request
